spec/ics-3-bidirectional-connection-semantics/README.md

```diff
@@ -0,0 +1,358 @@
+
+This ICS defines four subprotocols: opening handshake, header tracking, closing handshake, and closing by equivocation.
+
+##### Opening Handshake
+
+The opening handshake subprotocol serves to initialize roots of trust for two chains on each other and negotiate an agreeable connection version.
+
+The opening handshake defines four datagrams: *ConnOpenInit*, *ConnOpenTry*, *ConnOpenAck*, and *ConnOpenConfirm*.
+
+A correct protocol execution flows as follows:
+
+| Initiator | Datagram          | Chain |
+| --------- | ----------------- | ----- |
+| Actor     | `ConnOpenInit`    | A     |
+| Relayer   | `ConnOpenTry`     | B     |
+| Relayer   | `ConnOpenAck`     | A     |
+| Relayer   | `ConnOpenConfirm` | B     |
+
+At the end of an opening handshake between two chains implementing the subprotocol, the following properties hold:
+- Each chain has each other's correct root-of-trust as originally specified by the initiating actor.
+- The chains have agreed to a shared connection version.
+- Each chain has knowledge of and has agreed to its identifier on the other chain.
+
+This subprotocol need not be permissioned, modulo anti-spam measures.
+
+*ConnOpenInit* initializes a connection attempt on chain A.
+
+```golang
+type ConnOpenInit struct {
+  // Identifier to use for connection on chain A
+  Identifier  identifier
+  // Desired identifier to use for connection on chain B
+  Identifier  desiredCounterpartyIdentifier
+  // Desired version for connection
+  Version     desiredVersion
+  // Light client identifier for chain B
+  Identifier lightClientIdentifier
+}
+```
+
+```coffeescript
+function handleConnOpenInit(identifier, desiredVersion, desiredCounterpartyIdentifier, lightClientIdentifier)
+  assert(Get(identifier) == null)
+  state = INIT
+  Set(identifier, (state, desiredVersion, desiredCounterpartyIdentifier, lightClientIdentifier))
+```
+
+*ConnOpenTry* relays notice of a connection attempt on chain A to chain B.
+
+```golang
+type ConnOpenTry struct {
+  // Desired identifier to use for connection on chain B
+  Identifier        desiredIdentifier
+  // Identifier for connection on chain A
+  Identifier        counterpartyIdentifier
+  // Desired version for connection
+  Version           desiredVersion
+  // Light client identifier for chain B on A
+  Identifier        counterpartyLightClientIdentifier
+  // Light client identifier for chain A on B
+  Identifier        lightClientIdentifier
+  // Proof of stored INIT state on chain A
+  AccumulatorProof  proofInit
+}
+```
+
+```coffeescript
+function handleConnOpenTry(desiredIdentifier, counterpartyIdentifier, desiredVersion, counterpartyLightClientIdentifier, lightClientIdentifier, proofInit)
```

Comment:
Should all of these functions provide a specify a timeout handler if the next step of protocol isn't executed within N blocks?

Reply:
Yes, it should, and for safety I think the datagrams need to contain the block number as well. 

I want to work on channel definitions quickly first but then I'll add this.

Reply:
See https://github.com/cosmos/ics/issues/86

Reply:
Timeout handler specified, also see the state machine diagram.